walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,86 @@
+    
+    ArrayList<TimeRange> availableTimesForAllAttendees = 
+      new ArrayList<TimeRange>(availableTimesForMandatoryAttendeesOnly);
+    availableTimesForAllAttendees.retainAll(availableTimesForOptionalAttendeesOnly);
+
+    if(availableTimesForAllAttendees.size() > 0) return availableTimesForAllAttendees;
+
+    if(request.getAttendees().size() == 0 && request.getOptionalAttendees().size() > 0) {
+      return availableTimesForOptionalAttendeesOnly;
+    }
+
+    return availableTimesForMandatoryAttendeesOnly;
+    
+  }
+
+
+  private ArrayList<TimeRange> getAvailableTimes(Collection<Event> events, MeetingRequest request, Collection<String> attendees) {
+
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) return new ArrayList<TimeRange>();
+
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    availableTimes.add(TimeRange.WHOLE_DAY);
+
+    for(Event event : events) {
+
+      boolean eventHasMeetingAttendee = false;
+      for(String eventAttendee : event.getAttendees()) {
+        if(attendees.contains(eventAttendee)) {
+          eventHasMeetingAttendee = true;
+          break;
+        }
+      }
+      if(!eventHasMeetingAttendee) continue;
+
+      TimeRange eventTime = event.getWhen();
+      for(int i = 0; i < availableTimes.size(); i++) {
+        // if contains else if overlap
+        TimeRange currentAvailableTime = availableTimes.get(i);
+        if(availableTimes.get(i).contains(eventTime)) {
+          //break up interval into two
+          TimeRange splitLower = TimeRange.fromStartEnd(currentAvailableTime.start(), eventTime.start(), false);
+          TimeRange splitUpper = TimeRange.fromStartEnd(eventTime.end(), currentAvailableTime.end(), false);
+          availableTimes.remove(i);
+          if(splitUpper.duration() >= request.getDuration()) availableTimes.add(i, splitUpper);
+          if(splitLower.duration() >= request.getDuration()) availableTimes.add(i, splitLower);
+        }
+        else if(availableTimes.get(i).overlaps(eventTime)) {
+          // check starts and ends and split apppropriately
+          if(eventTime.start() < currentAvailableTime.start()) {
+            TimeRange modifiedAvailableTime = TimeRange.fromStartEnd(eventTime.end(), currentAvailableTime.end(), false);
+            if(modifiedAvailableTime.duration() >= request.getDuration()) {
+              availableTimes.set(i, modifiedAvailableTime);
+            }
+            else {
+              availableTimes.remove(i);
+            }
+          }
+          else if(eventTime.start() > currentAvailableTime.start()) {
+            TimeRange modifiedAvailableTime = TimeRange.fromStartEnd(currentAvailableTime.start(), eventTime.start(), false);
+            if(modifiedAvailableTime.duration() >= request.getDuration()) {
+              availableTimes.set(i, modifiedAvailableTime);
+            }
+            else {
+              availableTimes.remove(i);
+            }
+          }
+        }
```

Comment:
I would encourage you to factor this into one or more subroutines.

There's a few things you're doing here:

1. Determining if intervals overlap and then either splitting the available time, or getting the half of the available time that isn't in the event time. You're essentially doing a [set difference](https://www.probabilitycourse.com/chapter1/1_2_2_set_operations.php) with time intervals so see if that can be implemented in its own function.
2. Checking if the modified time(s) are long enough.
3. Replacing the existing available time with one or two new modified times.